docs/upgrading/remix.md

```diff
@@ -5,22 +5,125 @@ hidden: true
+| `build`     | `remix vite:build`                  | ➡️  | `react-router build`                       |
+| `start`     | `remix-serve build/server/index.js` | ➡️  | `react-router-serve build/server/index.js` |
+| `typecheck` | `tsc`                               | ➡️  | `react-router typegen && tsc`              |
+
+### Step 4 - Rename plugin in `vite.config`
+
+Update the import and rename the plugin in your `vite.config.ts`:
+
+```diff
+-import { vitePlugin as remix } from "@remix-run/dev";
++import { reactRouter } from "@react-router/dev/vite";
+import { defineConfig } from "vite";
+import tsconfigPaths from "vite-tsconfig-paths";
+
+export default defineConfig({
+  plugins: [
+-   remix({
+-     future: {
+-       // all future flags adopted
+-     },
+-   }),
++   reactRouter(),
+    tsconfigPaths(),
+  ],
+});
+```
+
+### Step 5 - Add a `routes.ts` file
+
+In React Router v7 you define your routes using the [`app/routes.ts`][routing] file. For backwards-compatibility and for folks who prefer [file-based conventions][fs-routing], you can opt-into the same "flat routes" convention you are using in Remix v2 via the new `@react-router/fs-routes` package:
+
+```ts filename=app/routes.ts
+import { type RouteConfig } from "@react-router/dev/routes";
+import { flatRoutes } from "@react-router/fs-routes";
+
+export const routes: RouteConfig = flatRoutes();
+```
+
+### Step 6 - Rename components in entry files
+
+If you have an `entry.server.tsx` and/or an `entry.client.tsx` file in your application, you will need to rename the main components in this files:
+
+| Entry File         | Remix v2 Component |     | React Router v7 Component |
+| ------------------ | ------------------ | --- | ------------------------- |
+| `entry.server.tsx` | `<RemixServer>`    | ➡️  | `<ServerRouter>`          |
+| `entry.client.stx` | `<RemixBrowser>`   | ➡️  | `<HydratedRouter>`        |
+
+## Known Prerelease Issues
+
+### Typesafety
+
+We have introduced some major changes to improve the type story in v7, but we're still working on making sure the path to adopt them is as smooth as possible prior to a stable v7 release. You can read more about the new type story in the [v7 draft release notes][v7-changelog-types] and if it's not a huge lift, your best bet for types in v7 is to migrate to that approach across the board.
+
+For the time being we don't have a great story to _incrementally_ migrate data types to the v7 prerelease. We never brought the generics on data APIs (`useLoaderData`, `useFetcher`, `Await`, etc.) over from Remix to React Router because we knew that we could do better than what Remix v2 had, and we wanted to ensure that we didn't ship APIs in React Router just to yank them out. Now that we have a better idea of the type story in React outer v7, we're better able to see what the migration path look like and we plan on shipping improvements in this area in an upcoming v7 prerelease.
```

Comment:
```suggestion
For the time being we don't have a great story to _incrementally_ migrate data types to the v7 prerelease. We never brought the generics on data APIs (`useLoaderData`, `useFetcher`, `Await`, etc.) over from Remix to React Router because we knew that we could do better than what Remix v2 had, and we wanted to ensure that we didn't ship APIs in React Router just to yank them out. Now that we have a better idea of the type story in React Router v7, we're better able to see what the migration path looks like and we plan on shipping improvements in this area in an upcoming v7 prerelease.
```